### PR TITLE
Adding support for executing transactions in scripting mode

### PR DESCRIPTION
**Pull request by falvaradorodriguez**
Closes https://github.com/safe-global/safe-cli/issues/36

Points of interest:

- Added `send_ether`, `send-erc20`, `send-erc721`, `send-custom` and `tx-builder` commands. The first commands allow  to execute transactions and the second, by importing a file generated by the tx_builder, can execute tx batches using the MultiSendCallOnly contract.
- ~~The logic of the `tx_builder_file_decoder` is developed based on what I know is done in the front end [tx_builder](https://github.com/safe-global/safe-react-apps/blob/development/apps/tx-builder/src/store/transactionLibraryContext.tsx).~~ Added here: https://github.com/safe-global/safe-cli/pull/426
- A ~~`script_mode`~~ `no_input ` property is added to `safe_operator` to raise exceptions that terminate execution instead of just displaying the error message (prompt mode).
- Updated the `Readme` file with the examples of the added commands.

### CODE REVIEW

File: src/safe_cli/operators/safe_operator.py [hunk 1]
```diff
@@ -182,6 +186,7 @@ def __init__(self, address: ChecksumAddress, node_url: str):
             True  # Require all signatures to be present to send a tx
         )
         self.hw_wallet_manager = get_hw_wallet_manager()
+        self.script_mode = script_mode  # Disable prompt dialogs
```
Comment (fmrsabino) [inline src/safe_cli/operators/safe_operator.py:189]: I think `script_mode` can be confusing as it might suggest that we offering ways to load and scripts made specifically for the `safe-cli`. The way I see it is that this is to support calls from scripts where there won't be any user input. If that's the case maybe we can rename this to `no_input` what do you think?

This feature can be enabled via `--no-input` and would be disabled for REPL by default.

File: src/safe_cli/operators/safe_operator.py [hunk 2]
```diff
@@ -936,7 +943,7 @@ def execute_safe_transaction(self, safe_tx: SafeTx):
             else:
                 call_result = safe_tx.call(self.hw_wallet_manager.sender.address)
             print_formatted_text(HTML(f"Result: <ansigreen>{call_result}</ansigreen>"))
-            if yes_or_no_question("Do you want to execute tx " + str(safe_tx)):
+            if self._is_confirmed_transaction_execution(safe_tx):
```
Comment (fmrsabino) [inline src/safe_cli/operators/safe_operator.py:946]: nitpick: I find that extracting this to a function to be more confusing. I think it's clearer if the expression is in the `if` itself:

```python
if self.script_mode or yes_or_no_question(f"Do you want to execute tx {safe_tx}"):
```

Comment (fmrsabino) [reply src/safe_cli/operators/safe_operator.py:946]: An alternative would be to try to embed the `script_mode` directly in the `yes_or_no_question` as then we would cover all use cases. If we miss one `yes_or_no_question` this might block the script.

File: src/safe_cli/tx_builder/tx_builder_file_decoder.py [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+    value: int
+    data: str
+
+    def __str__(self):
+        return f"id={self.id} to={self.to} value={self.value} data={self.data}"
+
+
+def convert_to_proposed_transactions(
+    batch_file: Dict[str, Any]
+) -> List[SafeProposedTx]:
```
Comment (fmrsabino) [inline src/safe_cli/tx_builder/tx_builder_file_decoder.py:224]: Is this new functionality? If so I suggest moving it into a different Pull Request. The scope of this PR is already quite big to introduce new functionality with it in my opinion (and reverting the CLI changes proposed here should not necessarily revert this new feature).

**Review by fmrsabino — CHANGES_REQUESTED** [review]

File: src/safe_cli/safe_cli.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+    def get_command(self) -> str:
+        return self.session.prompt(
+            self.get_prompt_text,
+            auto_suggest=AutoSuggestFromHistory(),
+            bottom_toolbar=self.get_bottom_toolbar,
+            lexer=PygmentsLexer(SafeLexer),
+            completer=SafeCompleter(),
+        )
+
+    def loop(self):
```
[the remark this answers is not in the document]
Comment (falvaradorodriguez) [reply src/safe_cli/safe_cli.py:108]: When starts "print_startup_info" method, a message is being displayed that shows:

"Version: 1.2.1
Loading Safe information..." 

and then waits for the response with the Safe information. Do you think we should change something here?

File: src/safe_cli/main.py [hunk 5]
```diff
@@ -1,193 +1,405 @@
...
+    epilog="Commands available in unattended mode:\n\n\n\n"
+    + "\n\n".join(
+        [
+            f"  {get_command_name(command)}"
+            for command in get_command(app).commands.keys()
+        ]
+    )
+    + "\n\n\n\nUse the --help option of each command to see the usage options.",
+)
+def default_attended_mode(
```
[the remark this answers is not in the document]
Comment (falvaradorodriguez) [reply src/safe_cli/main.py:362]: It makes sense and I agree. IMO we should do this in a new task, as the purpose of this task was to create the unattended commands and the import of the tx-builder json.
What do you think about doing it in a new task so as not to make this PR bigger?

File: src/safe_cli/safe_cli.py [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+
+from safe_cli.operators import (
+    SafeCliTerminationException,
+    SafeOperator,
+    SafeServiceNotAvailable,
+    SafeTxServiceOperator,
+)
+from safe_cli.prompt_parser import PromptParser
+from safe_cli.safe_completer import SafeCompleter
+from safe_cli.safe_lexer import SafeLexer
```
[the remark this answers is not in the document]
Comment (falvaradorodriguez) [reply src/safe_cli/safe_cli.py]: Done [here](https://github.com/safe-global/safe-cli/pull/419/commits/1e1d3548bd4c283e2de19261d3936d82d86bf698)

File: src/safe_cli/main.py [hunk 7]
```diff
@@ -1,193 +1,405 @@
...
+from safe_cli import VERSION
+from safe_cli.argparse_validators import check_hex_str
+from safe_cli.operators import SafeOperator
+from safe_cli.safe_cli import SafeCli
+from safe_cli.tx_builder.tx_builder_file_decoder import convert_to_proposed_transactions
+from safe_cli.typer_validators import (
+    check_ethereum_address,
+    check_private_keys,
+    parse_checksum_address,
+    parse_hex_str,
```
[the remark this answers is not in the document]
Comment (falvaradorodriguez) [reply src/safe_cli/main.py]: Done [here](https://github.com/safe-global/safe-cli/pull/419/commits/1e1d3548bd4c283e2de19261d3936d82d86bf698)

File: src/safe_cli/main.py [hunk 8]
```diff
@@ -1,193 +1,405 @@
...
-            f"{self.safe_operator.safe_cli_info}</style></b>"
-        )
+@app.command()
+def send_ether(
+    safe_address: Annotated[
+        ChecksumAddress,
+        typer.Argument(
+            help="The address of the Safe.",
+            callback=check_ethereum_address,
+            parser=parse_checksum_address,
```
[the remark this answers is not in the document]
Comment (falvaradorodriguez) [reply src/safe_cli/main.py]: Makes sense, by default takes the name of the parse function. I add a [CustomParser](https://typer.tiangolo.com/tutorial/parameter-types/custom-types/) to improve it. Done [here](https://github.com/safe-global/safe-cli/pull/419/commits/1e1d3548bd4c283e2de19261d3936d82d86bf698)

File: src/safe_cli/operators/safe_operator.py (hunk 1 above)
Comment (falvaradorodriguez) [reply src/safe_cli/operators/safe_operator.py]: Done [here](https://github.com/safe-global/safe-cli/pull/419/commits/1e1d3548bd4c283e2de19261d3936d82d86bf698). Do you see it necessary to add the `--no-input` option to all methods? It may not aport a lot, as internally it is easy to differentiate when the attended mode or unattended mode is used.

File: src/safe_cli/operators/safe_operator.py (hunk 2 above)
Comment (falvaradorodriguez) [reply src/safe_cli/operators/safe_operator.py]: Done [here](https://github.com/safe-global/safe-cli/pull/419/commits/1e1d3548bd4c283e2de19261d3936d82d86bf698).

File: src/safe_cli/tx_builder/tx_builder_file_decoder.py (hunk 3 above)
Comment (falvaradorodriguez) [reply src/safe_cli/tx_builder/tx_builder_file_decoder.py]: I have finally reverted the changes related to tx-builder. I have created a branch from this one incorporating them. The PR of the new branch is: https://github.com/safe-global/safe-cli/pull/426

File: src/safe_cli/safe_cli.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (fmrsabino) [reply src/safe_cli/safe_cli.py:107]: As discussed, it seems that there is no delay except in one situation: if `--get-safes-from-owner` is enabled then we need to remotely connect to another service which can take some time.

We should consider providing some type of feedback to the user where we indicate that it's fetching data (this happens before the call to `print_startup_info` https://github.com/safe-global/safe-cli/blob/ded64e7a800979f5394137648871f3376d67b269/src/safe_cli/main.py#L328

It can be done on a separate PR if we decide doing this 👍

File: src/safe_cli/main.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (fmrsabino) [reply src/safe_cli/main.py:299]: > It makes sense and I agree. IMO we should do this in a new task, as the purpose of this task was to create the unattended commands and the import of the tx-builder json. What do you think about doing it in a new task so as not to make this PR bigger?

I agree. Let's tackle this in a new task 🙂

File: src/safe_cli/operators/safe_operator.py (hunk 1 above)
Comment (fmrsabino) [reply src/safe_cli/operators/safe_operator.py]: > Done [here](https://github.com/safe-global/safe-cli/pull/419/commits/1e1d3548bd4c283e2de19261d3936d82d86bf698). Do you see it necessary to add the `--no-input` option to all methods? It may not aport a lot, as internally it is easy to differentiate when the attended mode or unattended mode is used.

I'd start with adding to the methods that require some input and try to find a way to have it for all the methods. The main reason being that it's an option that makes sense for any command even if there's no user prompt for some of them. But this makes it learning the CLI easier as the the user wouldn't need to know where `--no-input` is valid or not.

If there's a way in Typer to have "common" configs then we could use that.

**Review by fmrsabino — COMMENTED** [review]
Great job 👏 

I was considered additionally that the purpose of this PR, in addition to the "attended-mode" was to support scripting. In that case I think having a `--no-input` flag or similar should be added (initially to the commands that require it but in the future for it to be a global parameter).

File: src/safe_cli/safe_cli.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (falvaradorodriguez) [reply src/safe_cli/safe_cli.py:103]: Done [here](https://github.com/safe-global/safe-cli/pull/419/commits/7f61716537adc770043c238b88587ea86260729f)

File: src/safe_cli/main.py [hunk 9]
```diff
@@ -1,193 +1,342 @@
...
 
 
-if __name__ == "__main__":
-    main()
+def main():
+    # By default, the attended mode is initialised. Otherwise, the required command must be specified.
+    if len(sys.argv) == 1 or sys.argv[1] not in [
+        get_command_name(key) for key in get_command(app).commands.keys()
+    ]:
+        sys.argv.insert(1, "attended-mode")
```
[the remark this answers is not in the document]
Comment (falvaradorodriguez) [reply src/safe_cli/main.py:411]: Good catch! I think I solve it this [way](https://github.com/safe-global/safe-cli/pull/419/commits/4d93c9f8bcbd291bdb30d6a6aefddc6eaf76aa5e)

File: src/safe_cli/main.py [hunk 10]
```diff
@@ -135,9 +163,17 @@ def send_erc20(
             show_default=False,
         ),
     ] = None,
+    interactive: Annotated[
```
[the remark this answers is not in the document]
Comment (fmrsabino) [reply src/safe_cli/main.py:166]: Also suggested a new `help` message. Let me know what you think! 🙂

**Review by fmrsabino — APPROVED** [review]
Great work 👏 👏 👏
